### PR TITLE
Alow Mutating Entry keys and iterator keys

### LINKED ISSUE #334: Expanding mut key API
`toml_edit` presents `Table` as a mapping between `Key` and `Value` with `KeyMut` for providing access to the subset of key functionality that the user can modify.  Internally, this is stored as `IndexMap<String, (Key, Value)>`, where `Key` contains a `String`, among other data.

`Key.get().clone()` to get the `String` key is showing up in my benchmarks in parsing a real-world `Cargo.toml` file.  I'm exploring my options for avoiding that

 I hadn't noticed `MutableKeys` before and it seems like a good start by allowing me to instead do `IndexMap<Key, Value>`.  Some parts I'm missing for maintaining my `KeyMut` access for users:
- `MutableKeys::iter_full_mut2`
- `MutableOccupiedEntrykeys::get_full_mut2`

There may be more as I've not fully gone through my compilation errors.  I'm more wondering if this direction is viable before I continue down this path.

I'd be willing to contribute these if there is interest.

### PR DESCRIPTION
**Pull request by epage**
See toml-rs/toml#765 for an example use case

Fixes #334

### CODE REVIEW

**Comment by cuviper** [comment]
> See [toml-rs/toml#765](https://github.com/toml-rs/toml/pull/765) for an example use case

Any performance numbers to share from that change?

**Comment by epage** [comment]
Machine has been pretty jittery though,. so using callgrind I get
- Before: 8,879,894
- After: 8,678,207

This also cleaned up the code a lot.

File: src/map/mutable.rs [hunk 1]
```diff
@@ -80,8 +90,77 @@ where
...
+    fn key_mut(&mut self) -> &mut Self::Key;
+}
+
+/// Opt-in mutable access to [`Entry`] keys.
+///
+/// See [`MutableEntryKey`] for more information.
+impl<K, V> MutableEntryKey for Entry<'_, K, V> {
+    type Key = K;
+
+    /// Gets a nutable reference to the entry's key, either within the map if occupied,
```
Comment (cuviper) [inline src/map/mutable.rs:118]:
```suggestion
    /// Gets a mutable reference to the entry's key, either within the map if occupied,
```

**Comment by cuviper** [comment]
Thanks! I've published 2.3.0 with these changes.

**Pull request merged**